🐛  update CoreDNS ClusterRole for Kubernetes >= 1.22 and CoreDNS >= 1.8.1

sbueringer opened the pull request:
Signed-off-by: Stefan Büringer buringerst@vmware.com

<!-- please add a icon to the title of this PR (see https://sigs.k8s.io/cluster-api/VERSIONING.md), and delete this line and similar ones -->
<!-- the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other) -->

**What this PR does / why we need it**:
See issue for details

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #4927

sbueringer commented:
/test pull-cluster-api-e2e-full-main
/test pull-cluster-api-e2e-workload-upgrade-1-21-latest-main

vincepri commented on controlplane/kubeadm/internal/workload_cluster_coredns.go:
```diff
@@ -218,14 +251,52 @@ func (w *Workload) updateCoreDNSDeployment(ctx context.Context, info *coreDNSInf
+			return fmt.Errorf("failed to get ClusterRole %q", coreDNSClusterRoleName)
+		}
+
+		updatedClusterRole := coreDNSClusterRole.DeepCopy()
+		updatedClusterRole.ResourceVersion = currentClusterRole.ResourceVersion
+
+		if !reflect.DeepEqual(currentClusterRole, updatedClusterRole) {
+			if err := w.Client.Update(ctx, updatedClusterRole); err != nil {
+				return errors.Wrapf(err, "failed to update ClusterRole %q", coreDNSClusterRoleName)
+			}
+		}
```
We should probably change this code to apply the coreDNSClusterRole `Rules` field on the `currentClusterRole` and not forcing the `ResourceVersion` which effectively gets around the Update strong consistency requirement

vincepri commented on controlplane/kubeadm/internal/workload_cluster_coredns.go:
```diff
@@ -218,14 +251,52 @@ func (w *Workload) updateCoreDNSDeployment(ctx context.Context, info *coreDNSInf
+
+	return retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		var currentClusterRole rbacv1.ClusterRole
+		err = w.Client.Get(ctx, ctrlclient.ObjectKey{Name: coreDNSClusterRoleName, Namespace: metav1.NamespaceSystem}, &currentClusterRole)
+		if err != nil {
```
```suggestion
		if err := w.Client.Get(ctx, ctrlclient.ObjectKey{Name: coreDNSClusterRoleName, Namespace: metav1.NamespaceSystem}, &currentClusterRole)
```

vincepri replied:
Also put the ObjectKey outside of the Retry?

vincepri commented on controlplane/kubeadm/internal/workload_cluster_coredns.go:
```diff
@@ -218,14 +251,52 @@ func (w *Workload) updateCoreDNSDeployment(ctx context.Context, info *coreDNSInf
+// To support Kubernetes clusters >= 1.22 (which have been initialized with kubeadm < 1.22) with CoreDNS versions >= 1.8.1
+// we have to update the ClusterRole accordingly.
+func (w *Workload) updateCoreDNSClusterRole(ctx context.Context, kubernetesVersion semver.Version, info *coreDNSInfo) error {
+	// do nothing for Kubernetes < 1.22
```
```suggestion
	// Do nothing for Kubernetes < 1.22.
```

vincepri commented on controlplane/kubeadm/internal/workload_cluster_coredns.go:
```diff
@@ -218,14 +251,52 @@ func (w *Workload) updateCoreDNSDeployment(ctx context.Context, info *coreDNSInf
+		return nil
+	}
+
+	// do nothing for CoreDNS < 1.8.1
```
```suggestion
	// Do nothing for CoreDNS < 1.8.1.
```

vincepri commented on controlplane/kubeadm/internal/workload_cluster_coredns.go:
```diff
@@ -19,35 +19,65 @@ package internal
 
+var (
+	// Source: https://github.com/kubernetes/kubernetes/blob/v1.22.0-beta.1/cmd/kubeadm/app/phases/addons/dns/manifests.go#L178-L207
+	coreDNSClusterRole = &rbacv1.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      coreDNSClusterRoleName,
+			Namespace: metav1.NamespaceSystem,
+		},
+		Rules: []rbacv1.PolicyRule{
```
Can we simplify to only list the rules that need to be updated instead of the whole ClusterRole object? Something like `coreDNS181PolicyRules`

sbueringer commented:
/test pull-cluster-api-e2e-full-main
/test pull-cluster-api-e2e-workload-upgrade-1-21-latest-main

@vincepri ptal

vincepri commented on controlplane/kubeadm/internal/workload_cluster_coredns.go:
```diff
@@ -218,14 +245,51 @@ func (w *Workload) updateCoreDNSDeployment(ctx context.Context, info *coreDNSInf
+
+	key := ctrlclient.ObjectKey{Name: coreDNSClusterRoleName, Namespace: metav1.NamespaceSystem}
+	return retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		var currentClusterRole rbacv1.ClusterRole
+		if err = w.Client.Get(ctx, key, &currentClusterRole); err != nil {
```
```suggestion
		if err := w.Client.Get(ctx, key, &currentClusterRole); err != nil {
```

vincepri replied:
```suggestion
	currentClusterRole := &rbacv1.ClusterRole{]
		if err = w.Client.Get(ctx, key, currentClusterRole); err != nil {
```

vincepri commented on controlplane/kubeadm/internal/workload_cluster_coredns.go:
```diff
@@ -218,14 +245,51 @@ func (w *Workload) updateCoreDNSDeployment(ctx context.Context, info *coreDNSInf
+		}
+
+		if !reflect.DeepEqual(currentClusterRole.Rules, coreDNS181PolicyRules) {
+			updatedClusterRole := currentClusterRole.DeepCopy()
```
Do we need to copy here?

sbueringer replied:
I think I was confusing it with something along the lines of "we have to deep copy otherwise we will change the object in the cache which could lead to strange bugs". But I assume that's not an issue here as we're using our own object.

sbueringer commented:
/test pull-cluster-api-e2e-full-main
/test pull-cluster-api-e2e-workload-upgrade-1-21-latest-main

sbueringer commented:
/test pull-cluster-api-e2e-workload-upgrade-1-21-latest-main
I think it's an unrelated error. Also so that one in the testgrid before. Otherwise I'll take a closer look tomorrow.

sbueringer commented:
/test pull-cluster-api-e2e-full-main
/test pull-cluster-api-e2e-workload-upgrade-1-21-latest-main

sbueringer commented:
Okay looks like I got the tests green now. I also adjusted our pull image code (which was merged yesterday). So this PR now also fixes all our CI jobs which are using locally build kindest/node images.

I'll squash before merge.

sbueringer commented:
/test pull-cluster-api-e2e-full-main
/test pull-cluster-api-e2e-workload-upgrade-1-21-latest-main

sbueringer commented:
/retetst

sbueringer commented:
/retest

stmcginnis commented:
/lgtm

sbueringer commented:
/retest

vincepri commented on controlplane/kubeadm/internal/workload_cluster_coredns.go:
```diff
@@ -218,14 +245,50 @@ func (w *Workload) updateCoreDNSDeployment(ctx context.Context, info *coreDNSInf
+			return fmt.Errorf("failed to get ClusterRole %q", coreDNSClusterRoleName)
+		}
+
+		if !reflect.DeepEqual(currentClusterRole.Rules, coreDNS181PolicyRules) {
```
Given that these are slices we might have to either sort them, or do a comparison one by one. See https://play.golang.org/p/CJaurc7dzdz

vincepri replied:
/hold

sbueringer replied:
Yeah agree. I'll take a closer look tomorrow.

sbueringer replied:
I pushed a new commit with an implementation (I'll squash after we got consensus).

I implemented it by building a map with all policies and comparing them with deep equal. While this probably needs a bit more memory (which should be negligible), I think it should be faster then recursively sorting (basically) nested arrays (although that also shouldn't be that important as we're running this code only when updating CoreDNS).

sbueringer commented:
/test pull-cluster-api-e2e-full-main
/test pull-cluster-api-e2e-workload-upgrade-1-21-latest-main

sbueringer commented:
/test pull-cluster-api-e2e-full-main
/test pull-cluster-api-e2e-workload-upgrade-1-21-latest-main

sbueringer commented:
/assign @fabriziopandini 
(if you have some time, this PR fixes all our upgrade tests again)

vincepri commented on controlplane/kubeadm/internal/workload_cluster_coredns.go:
```diff
@@ -218,14 +245,76 @@ func (w *Workload) updateCoreDNSDeployment(ctx context.Context, info *coreDNSInf
+
+	key := ctrlclient.ObjectKey{Name: coreDNSClusterRoleName, Namespace: metav1.NamespaceSystem}
+	return retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		currentClusterRole := &rbacv1.ClusterRole{}
+		if err := w.Client.Get(ctx, key, currentClusterRole); err != nil {
+			return fmt.Errorf("failed to get ClusterRole %q", coreDNSClusterRoleName)
+		}
+
+		if !semanticDeepEqualPolicyRules(currentClusterRole.Rules, coreDNS181PolicyRules) {
+			currentClusterRole.Rules = coreDNS181PolicyRules
+			if err := w.Client.Update(ctx, currentClusterRole); err != nil {
+				return errors.Wrapf(err, "failed to update ClusterRole %q", coreDNSClusterRoleName)
+			}
+		}
```
Instead of using update, should we use `Patch` by diffing the before/after objects and apply it using strong consistency?

sbueringer replied:
We have a lot of other updates here, but I can implement it.

Looks like our patchHelper is  onlyable to patch: spec, metadata and status. As ClusterRole doesn't has a spec I would have to implement the diff by myself. I'll do it if you think it's worth it.

vincepri reviewed (approved):
/lgtm

vincepri commented:
/assign @CecileRobertMichon

stmcginnis commented:
/lgtm

Updates look good, and really nice test coverage on the policy comparison.

sbueringer commented:
Squashed and rebased onto current main branch

/test pull-cluster-api-e2e-full-main
/test pull-cluster-api-e2e-workload-upgrade-1-21-latest-main

vincepri commented:
/lgtm

fabriziopandini commented:
@sbueringer thanks for tackling this!
/lgtm
/approve

sbueringer commented:
@fabriziopandini There's still a hold from: https://github.com/kubernetes-sigs/cluster-api/pull/4937#discussion_r670516666 should we unhold and merge?

vincepri commented:
/hold cancel
